pkg/controller/controller_test.go

```diff
@@ -2781,3 +2747,190 @@ func TestProvisionFromPVC(t *testing.T) {
+				Parameters:    map[string]string{"fstype": "ext3"},
+				ReclaimPolicy: &deletePolicy,
+			},
+			PVName: "test-name",
+			PVC: &v1.PersistentVolumeClaim{
+				ObjectMeta: metav1.ObjectMeta{
+					UID:         "testid",
+					Name:        "fake-pvc",
+					Annotations: tc.annotation,
+				},
+				Spec: v1.PersistentVolumeClaimSpec{
+					Selector: nil, // Provisioner doesn't support selector
+					Resources: v1.ResourceRequirements{
+						Requests: v1.ResourceList{
+							v1.ResourceName(v1.ResourceStorage): resource.MustParse(strconv.FormatInt(requestBytes, 10)),
+						},
+					},
+				},
+			},
+		}
+
+		_, state, err := csiProvisioner.(controller.ProvisionerExt).ProvisionExt(volOpts)
+		if tc.expectErr && err == nil {
+			t.Errorf("test %q: Expected error, got none", k)
+		}
+		if !tc.expectErr && err != nil {
+			t.Errorf("test %q: got error: %v", k, err)
+		}
+
+		if controller.ProvisioningFinished != state {
+			t.Errorf("test %q: expected ProvisioningState %s, got %s", k, controller.ProvisioningFinished, state)
+		}
+
+	}
+}
+
+func TestDeleteMigration(t *testing.T) {
+	testCases := []struct {
+		name           string
+		pv             *v1.PersistentVolume
+		translateToCSI bool
+		expectErr      bool
+	}{
+		{
+			name: "normal migration",
+			// The PV could be any random in-tree plugin - it doesn't really
+			// matter here. We only care that the translation is called and the
+			// function will work after some CSI volume is created
+			pv:             &v1.PersistentVolume{},
+			translateToCSI: true,
+		},
+		{
+			name: "no migration",
+			pv: &v1.PersistentVolume{
+				Spec: v1.PersistentVolumeSpec{
+					PersistentVolumeSource: v1.PersistentVolumeSource{
+						CSI: &v1.CSIPersistentVolumeSource{
+							VolumeHandle: "mock-id",
+						},
+					},
+				},
+			},
+			translateToCSI: false,
+		},
+	}
+	for _, tc := range testCases {
+
+		// Set up test
+		tmpdir := tempDir(t)
+		defer os.RemoveAll(tmpdir)
+		mockController, driver, _, controllerServer, csiConn, err := createMockServer(t, tmpdir)
+		if err != nil {
+			t.Fatal(err)
+		}
+		mockTranslator := NewMockProvisionerCSITranslator(mockController)
+		defer mockController.Finish()
+		defer driver.Stop()
+		clientSet := fakeclientset.NewSimpleClientset()
+		pluginCaps, controllerCaps := provisionCapabilities()
+		csiProvisioner := NewCSIProvisioner(clientSet, 5*time.Second, "test-provisioner", "test", 5, csiConn.conn, nil, driverName, pluginCaps, controllerCaps, "", false, mockTranslator)
+
+		// Set expectations
+		mockTranslator.EXPECT().IsPVMigratable(gomock.Any()).Return(tc.translateToCSI).Times(1)
+		if tc.translateToCSI {
+			mockTranslator.EXPECT().TranslateInTreePVToCSI(gomock.Any()).Return(
+				&v1.PersistentVolume{
+					Spec: v1.PersistentVolumeSpec{
+						PersistentVolumeSource: v1.PersistentVolumeSource{
+							CSI: &v1.CSIPersistentVolumeSource{
+								VolumeHandle: "mock-id",
+							},
+						},
+					},
+				}, nil).Times(1)
+		}
```

Comment:
It's possible with mock/expect libraries like this to overfit on the current implementation. Is the behavior described by these expectations really invariant? It is invalid to call `IsPVMigratable` twice from `Delete`? And `TranslateInTreePVToCSI` should only be called if `IsPVMigratable` returns true?

Reply:
this is what I was worried about with this library in the first place. I definitely feel like I'm overfitting a little - IIUC that's kind of just how these libraries work... The intention is to not be perscriptive in the number of calls but rather to specify a specific return value so the test will continue to completion after doing translation.

> TranslateInTreePVToCSI should only be called if IsPVMigratable

This is invariant

> It is invalid to call IsPVMigratable twice from Delete

This is not.

Would it be more accurate to set this to `MinTimes`? Or do you have different guidance all-together. I'm not really too familiar with this type of testing so I don't really know how to avoid the overfitting problem since by nature we're asserting implementation details and not just relying on the input/output interfaces

Reply:
Updated so that its not overfitting as much on the implementation by using `AnyTimes` on implementation details (still similar semantics but now the main purpose of those mock calls is to set return value expectations).

We are doing extra strict checks on the "interfaces" of the function - inputs -> outputs (in this case both the function returns AND the inputs to the CSI calls)